File: services/streaming-service/internal/messaging/publisher.go
package messaging

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
	"github.com/rs/zerolog/log"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"

	"github.com/streamvault/streaming-service/internal/config"
	"github.com/streamvault/streaming-service/internal/telemetry"
)

type EncodingJob struct {
	JobID        string    `json:"job_id"`
	ContentID    string    `json:"content_id"`
	SourcePath   string    `json:"source_path"`
	SourceBucket string    `json:"source_bucket"`
	RequestedBy  string    `json:"requested_by"`
	CreatedAt    time.Time `json:"created_at"`
}

type WatchCompletedEvent struct {
	EventID       string             `json:"eventId"`
	EventType     string             `json:"eventType"`
	Timestamp     string             `json:"timestamp"`
	Source        string             `json:"source"`
	CorrelationID string             `json:"correlationId"`
	Data          WatchCompletedData `json:"data"`
}

type WatchCompletedData struct {
	UserID               string  `json:"userId"`
	ContentID            string  `json:"contentId"`
	CompletionPercentage float64 `json:"completionPercentage"`
	WatchedAt            string  `json:"watchedAt"`
}

type Publisher interface {
	PublishEncodingJob(ctx context.Context, job EncodingJob) error
	PublishWatchCompleted(ctx context.Context, event WatchCompletedEvent) error
	Close() error
}

type amqpPublisher struct {
	mu              sync.Mutex
	conn            *amqp.Connection
	channel         *amqp.Channel
	exchange        string
	routingKey      string
	watchExchange   string
	watchRoutingKey string
}

func NewPublisher(conn *amqp.Connection, cfg config.RabbitMQConfig) (Publisher, error) {
	ch, err := conn.Channel()
	if err != nil {
		return nil, fmt.Errorf("opening rabbitmq channel: %w", err)
	}

	log.Info().Str("exchange", cfg.Exchange).Str("routing_key", cfg.PublishRoutingKey).Msg("rabbitmq publisher ready")

	return &amqpPublisher{
		conn:            conn,
		channel:         ch,
		exchange:        cfg.Exchange,
		routingKey:      cfg.PublishRoutingKey,
		watchExchange:   cfg.WatchExchange,
		watchRoutingKey: cfg.WatchRoutingKey,
	}, nil
}

func (p *amqpPublisher) recreateChannel() error {
	if p.channel != nil {
		p.channel.Close()
	}
	ch, err := p.conn.Channel()
	if err != nil {
		return fmt.Errorf("recreating rabbitmq channel: %w", err)
	}
	p.channel = ch
	log.Info().Msg("rabbitmq publisher channel recreated")
	return nil
}

var publishTracer = otel.Tracer("streaming-service/messaging")

func (p *amqpPublisher) publish(ctx context.Context, exchange, routingKey string, messageID string, body []byte) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	ctx, span := publishTracer.Start(ctx, "rabbitmq.publish",
		trace.WithSpanKind(trace.SpanKindProducer),
		trace.WithAttributes(
			attribute.String("messaging.system", "rabbitmq"),
			attribute.String("messaging.destination.name", exchange),
			attribute.String("messaging.rabbitmq.routing_key", routingKey),
			attribute.String("messaging.message.id", messageID),
		),
	)
	defer span.End()

	headers := telemetry.InjectAMQP(ctx, nil)

	publishing := amqp.Publishing{
		ContentType:  "application/json",
		DeliveryMode: amqp.Persistent,
		MessageId:    messageID,
		Timestamp:    time.Now().UTC(),
		Headers:      headers,
		Body:         body,
	}

	err := p.channel.PublishWithContext(ctx, exchange, routingKey, false, false, publishing)
	if err != nil {
		log.Warn().Err(err).Msg("publish failed, attempting channel recreation")
		if recreateErr := p.recreateChannel(); recreateErr != nil {
			return fmt.Errorf("publish (channel recreation failed): %w", recreateErr)
		}
		err = p.channel.PublishWithContext(ctx, exchange, routingKey, false, false, publishing)
		if err != nil {
			return fmt.Errorf("publish (retry failed): %w", err)
		}
	}

	return nil
}

func (p *amqpPublisher) PublishEncodingJob(ctx context.Context, job EncodingJob) error {
	body, err := json.Marshal(job)
	if err != nil {
		return fmt.Errorf("marshalling encoding job: %w", err)
	}

	if err := p.publish(ctx, p.exchange, p.routingKey, job.JobID, body); err != nil {
		return fmt.Errorf("publishing encoding job: %w", err)
	}

	log.Info().Str("job_id", job.JobID).Str("content_id", job.ContentID).Msg("encoding job published")
	return nil
}

func (p *amqpPublisher) PublishWatchCompleted(ctx context.Context, event WatchCompletedEvent) error {
	body, err := json.Marshal(event)
	if err != nil {
		return fmt.Errorf("marshalling watch completed event: %w", err)
	}

	if err := p.publish(ctx, p.watchExchange, p.watchRoutingKey, event.EventID, body); err != nil {
		return fmt.Errorf("publishing watch completed event: %w", err)
	}

	log.Info().
		Str("userId", event.Data.UserID).
		Str("contentId", event.Data.ContentID).
		Float64("percentage", event.Data.CompletionPercentage).
		Msg("watch completed event published")
	return nil
}

func (p *amqpPublisher) Close() error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.channel != nil {
		return p.channel.Close()
	}
	return nil
}

File: services/encoding-service/src/messaging/publisher.rs
use async_trait::async_trait;
use lapin::options::BasicPublishOptions;
use lapin::types::FieldTable;
use lapin::{BasicProperties, Channel, Connection, ConnectionProperties};
use std::sync::Arc;
use tokio::sync::Mutex;
use tracing::{error, info, warn};

use crate::config::RabbitMQConfig;
use crate::error::{EncodingError, Result};
use crate::messaging::models::EncodingResult;
use crate::telemetry;

#[async_trait]
pub trait ResultPublisher: Send + Sync {
    async fn publish_completed(&self, result: EncodingResult) -> Result<()>;
    async fn publish_failed(&self, result: EncodingResult) -> Result<()>;
}

pub struct RabbitMQPublisher {
    connection: Connection,
    channel: Arc<Mutex<Channel>>,
    config: RabbitMQConfig,
}

impl RabbitMQPublisher {
    pub async fn new(config: RabbitMQConfig) -> Result<Self> {
        let connection =
            Connection::connect(&config.url, ConnectionProperties::default()).await?;
        let channel = connection.create_channel().await?;

        // Declare exchange to be safe (idempotent if already exists)
        channel
            .exchange_declare(
                &config.exchange,
                lapin::ExchangeKind::Topic,
                lapin::options::ExchangeDeclareOptions {
                    durable: true,
                    passive: false,
                    ..Default::default()
                },
                FieldTable::default(),
            )
            .await?;

        info!(exchange = %config.exchange, "rabbitmq publisher ready");

        Ok(Self {
            connection,
            channel: Arc::new(Mutex::new(channel)),
            config,
        })
    }

    #[tracing::instrument(
        skip_all,
        fields(
            messaging.system = "rabbitmq",
            messaging.destination.name = %self.config.exchange,
            messaging.rabbitmq.routing_key = %routing_key,
            messaging.message.id = %result.event_id,
            job.id = %result.job_id,
        ),
        name = "rabbitmq.publish"
    )]
    async fn publish(&self, result: &EncodingResult, routing_key: &str) -> Result<()> {
        let payload = serde_json::to_vec(result)?;
        let headers = telemetry::inject_context();
        let properties = BasicProperties::default()
            .with_content_type("application/json".into())
            .with_delivery_mode(2) // persistent
            .with_headers(headers);

        let publish_result = {
            let channel = self.channel.lock().await;
            channel
                .basic_publish(
                    &self.config.exchange,
                    routing_key,
                    BasicPublishOptions::default(),
                    &payload,
                    properties.clone(),
                )
                .await
        };

        match publish_result {
            Ok(confirm) => {
                confirm.await.map_err(|e| {
                    EncodingError::Internal(format!("publish confirm error: {}", e))
                })?;
                info!(
                    job_id = %result.job_id,
                    status = %result.status,
                    routing_key,
                    "published result"
                );
                Ok(())
            }
            Err(e) => {
                error!(error = ?e, "publish failed, recreating channel");
                self.recreate_channel().await?;

                let channel = self.channel.lock().await;
                channel
                    .basic_publish(
                        &self.config.exchange,
                        routing_key,
                        BasicPublishOptions::default(),
                        &payload,
                        properties,
                    )
                    .await?
                    .await
                    .map_err(|e| {
                        EncodingError::Internal(format!("publish confirm error: {}", e))
                    })?;

                info!(
                    job_id = %result.job_id,
                    status = %result.status,
                    "published result (after channel recreation)"
                );
                Ok(())
            }
        }
    }

    async fn recreate_channel(&self) -> Result<()> {
        let new_channel = self.connection.create_channel().await?;
        let mut channel = self.channel.lock().await;
        *channel = new_channel;
        info!("publisher channel recreated");
        Ok(())
    }

    pub fn connection(&self) -> &Connection {
        &self.connection
    }

    pub async fn close(&self) {
        let channel = self.channel.lock().await;
        if let Err(e) = channel.close(200, "shutdown").await {
            warn!(error = %e, "error closing publisher channel");
        }
        if let Err(e) = self.connection.close(200, "shutdown").await {
            warn!(error = %e, "error closing publisher connection");
        }
        info!("rabbitmq publisher connection closed");
    }
}

#[async_trait]
impl ResultPublisher for RabbitMQPublisher {
    async fn publish_completed(&self, result: EncodingResult) -> Result<()> {
        self.publish(&result, &self.config.completed_routing_key)
            .await
    }

    async fn publish_failed(&self, result: EncodingResult) -> Result<()> {
        self.publish(&result, &self.config.failed_routing_key).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::messaging::models::{EncodingOutput, EncodingResult};
    use std::sync::Mutex;

    /// Mock publisher that records all publish calls for assertion.
    struct TrackingPublisher {
        completed: Mutex<Vec<EncodingResult>>,
        failed: Mutex<Vec<EncodingResult>>,
    }

    impl TrackingPublisher {
        fn new() -> Self {
            Self {
                completed: Mutex::new(Vec::new()),
                failed: Mutex::new(Vec::new()),
            }
        }
    }

    #[async_trait]
    impl ResultPublisher for TrackingPublisher {
        async fn publish_completed(&self, result: EncodingResult) -> Result<()> {
            self.completed.lock().unwrap().push(result);
            Ok(())
        }
        async fn publish_failed(&self, result: EncodingResult) -> Result<()> {
            self.failed.lock().unwrap().push(result);
            Ok(())
        }
    }

    fn make_result(status: &str) -> EncodingResult {
        EncodingResult {
            event_id: "evt-001".into(),
            event_type: format!("encoding.job.{}", status),
            timestamp: "2024-01-15T10:35:00Z".into(),
            source: "encoding-service".into(),
            correlation_id: "job-001".into(),
            job_id: "job-001".into(),
            content_id: "movie-456".into(),
            status: status.into(),
            outputs: if status == "completed" {
                vec![EncodingOutput {
                    quality: "720p".into(),
                    width: 1280,
                    height: 720,
                    bitrate_kbps: 2800,
                    segment_count: 12,
                    playlist_path: "movie-456/720p/playlist.m3u8".into(),
                }]
            } else {
                vec![]
            },
            duration_seconds: 120,
            error_message: if status == "failed" {
                Some("codec not supported".into())
            } else {
                None
            },
            completed_at: "2024-01-15T10:35:00Z".into(),
        }
    }

    #[tokio::test]
    async fn publish_completed_records_result() {
        let publisher = TrackingPublisher::new();
        let result = make_result("completed");

        publisher.publish_completed(result).await.unwrap();

        let calls = publisher.completed.lock().unwrap();
        assert_eq!(calls.len(), 1);
        assert_eq!(calls[0].status, "completed");
        assert_eq!(calls[0].job_id, "job-001");
        assert_eq!(calls[0].content_id, "movie-456");
        assert_eq!(calls[0].outputs.len(), 1);
        assert_eq!(calls[0].outputs[0].quality, "720p");
        assert!(calls[0].error_message.is_none());
    }

    #[tokio::test]
    async fn publish_failed_records_result() {
        let publisher = TrackingPublisher::new();
        let result = make_result("failed");

        publisher.publish_failed(result).await.unwrap();

        let calls = publisher.failed.lock().unwrap();
        assert_eq!(calls.len(), 1);
        assert_eq!(calls[0].status, "failed");
        assert_eq!(calls[0].job_id, "job-001");
        assert_eq!(calls[0].outputs.len(), 0);
        assert_eq!(
            calls[0].error_message.as_deref(),
            Some("codec not supported")
        );
    }

    #[tokio::test]
    async fn publish_completed_and_failed_are_independent() {
        let publisher = TrackingPublisher::new();

        publisher
            .publish_completed(make_result("completed"))
            .await
            .unwrap();
        publisher
            .publish_failed(make_result("failed"))
            .await
            .unwrap();

        assert_eq!(publisher.completed.lock().unwrap().len(), 1);
        assert_eq!(publisher.failed.lock().unwrap().len(), 1);
    }

    #[tokio::test]
    async fn publish_result_event_envelope_fields() {
        let publisher = TrackingPublisher::new();
        let result = make_result("completed");

        publisher.publish_completed(result).await.unwrap();

        let calls = publisher.completed.lock().unwrap();
        assert_eq!(calls[0].event_id, "evt-001");
        assert_eq!(calls[0].event_type, "encoding.job.completed");
        assert_eq!(calls[0].source, "encoding-service");
        assert_eq!(calls[0].correlation_id, "job-001");
    }
}
